src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java

```diff
@@ -1049,31 +1049,43 @@ private void createFlattenedDependenciesDirect( List<Dependency> projectDependen
      * @throws ArtifactDescriptorException
      */
     private void createFlattenedDependenciesAll( List<Dependency> projectDependencies, List<Dependency> flattenedDependencies )
-            throws DependencyTreeBuilderException, ArtifactDescriptorException {
-        final Queue<DependencyNode> dependencyNodeLinkedList = new LinkedList<DependencyNode>() {};
+            throws DependencyTreeBuilderException, ArtifactDescriptorException
+    {
+        final Queue<DependencyNode> dependencyNodeLinkedList = new LinkedList<DependencyNode>();
         final Set<String> processedDependencies = new HashSet<>();
 
         final Artifact projectArtifact = this.project.getArtifact();
 
         final DependencyNode dependencyNode = this.dependencyTreeBuilder.buildDependencyTree(this.project,
                 this.localRepository, null);
 
-        dependencyNode.accept(new DependencyNodeVisitor() {
-            @Override public boolean visit(DependencyNode node) {
-                if (node.getArtifact().getGroupId().equals(projectArtifact.getGroupId()) && node.getArtifact().getArtifactId().equals(projectArtifact.getArtifactId())) {
+        dependencyNode.accept(new DependencyNodeVisitor()
+        {
+            @Override public boolean visit(DependencyNode node)
+            {
+                if (node.getArtifact().getGroupId().equals(projectArtifact.getGroupId()) && node.getArtifact().getArtifactId().equals(projectArtifact.getArtifactId()))
+                {
                     return true;
                 }
-                if (node.getState() != DependencyNode.INCLUDED) return true;
+                if (node.getState() != DependencyNode.INCLUDED) {
+                    return true;
+                }
+                if (node.getParent().getArtifact().isOptional())
+                {
+                    node.getArtifact().setOptional(true);
```

Comment:
my other concern is that we are mutating a node

Reply:
I can try to use HashSet to record the artifact we want to set as "optional" so we do not need to change the node.